### ISSUE TITLE
Sorting by credits or GPA over full DB is way slow

### ISSUE THREAD

**Issue by bjmckenz** [csufmse/student_information_system#159]
Aggregated data like credits, GPA, and class level is all calculated for each request. It's not denormalized and stored in the DB "like it would be in the real world".

This makes sorting on large sets (like all students) very slow at best.

Hope that's not a problem...

@jeffreicher would you mind marking #wontfix?

**Comment by liannadevries09** [comment]
@bjmckenz  - my bad. Didnt mean to close.. was see if it needed testing.

**Comment by bjmckenz** [comment]
Nope. markingas  wontfix for now.

**Comment by bjmckenz** [comment]
thinking of doing this; realized how it can be done well for our current db, although not what we'd do in the big world. This will then enable GPA and class level filtering.

**Comment by bjmckenz** [comment]
Testing:
* Verify that students load a bit faster
* verify that when a grade is entered (or changed or removed) the student's GPA and credits_earned match.

@oceancypress I kept thinking about your comment as to why not do this. I came to realize you were right and I was wrong. It was pretty easy.

It's not how we'd do it if the DB had triggers, but it's still decent. The upside is that it makes fetches and displays of GPA much faster. The downside is that updates are slower. This is only really noticeable on create_data (creating new db). But since this is an offline/one-time process, seems reasonable.

I chose to aggregate at the semester level first. Since we were talking a lot about semester gpa and that's a school thing (AP, for example) it seemed to make more sense. It also meant that an update to a section triggered a ~5 row fetch (classes in semester) then a ~10 row fetch (semesters for student) rather than a 50-row fetch (sections for student) every time.
